### PR TITLE
dialects: [experimental] Add a Stencil dialect

### PR DESCRIPTION
**Pull request by AntonLydike**

### CODE REVIEW

File: xdsl/dialects/experimental/stencil.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+@dataclass
+class IntOrUnknown(AttrConstraint):
+    length: int = 0
+
+    def verify(self, attr: Attribute) -> None:
+        if not isinstance(attr, ArrayAttr):
+            raise VerifyException(
+                f"Expected {ArrayAttr} attribute, but got {attr.name}.")
+
+        if len(attr.data) != self.length:
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/stencil.py:46]: How are we supposed to fix this? We don't care about the type inside the list, we just want the length? @superlopuh @math-fehr

[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/stencil.py:52]: Okay, I think "fixed" it with a cast to `ArrayAttr[Any]` but I don't really like it, it feels like a useless cast. Any ideas on how to improve this?

**Comment by AntonLydike** [comment]
@mesham I changed `ResultType.from_type` pretty drastically in [24e197b](https://github.com/xdslproject/xdsl/pull/469/commits/24e197beba4c6783a42dfba6ddc4e80a216b583b), I think that's the intended implementation, right? If not we can just revert this commit!

File: xdsl/dialects/experimental/stencil.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+            raise VerifyException(
+                f"Expected {ArrayAttr} attribute, but got {attr.name}.")
+        attr = cast(ArrayAttr[Any], attr)
+        if len(attr.data) != self.length:
+            raise VerifyException(
+                f"Expected array of length {self.length}, got {len(attr.data)}."
+            )
+
+
+@dataclass(frozen=True)
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/stencil.py:107]: @math-fehr How would we inherit from ParametrizedAttribute *and* MLIRType?

[the remark this answers is not in the document]
Comment (AntonLydike) [reply xdsl/dialects/experimental/stencil.py:107]: @math-fehr How would we inherit from ParametrizedAttribute *and* MLIRType?

File: xdsl/dialects/experimental/stencil.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+                                   IntegerType, IndexType, IntAttr, AnyFloat)
+from xdsl.ir import Operation, Dialect, MLIRType
+from xdsl.irdl import (irdl_attr_definition, irdl_op_definition, ParameterDef,
+                       AttrConstraint, Attribute, Region, VerifyException,
+                       AnyOf, Annotated, Operand, OpAttr, OpResult, VarOperand,
+                       VarOpResult, OptOpAttr, AttrSizedOperandSegments)
+
+
+@dataclass
+class IntOrUnknown(AttrConstraint):
```
Comment (webmiche) [inline xdsl/dialects/experimental/stencil.py:16]: Can you describe what this class models? The fact that a given Array has a specific length? Why the name then?

Comment (webmiche) [reply xdsl/dialects/experimental/stencil.py:16]: Wait isn't this the same as `ArrayLenght` below?

File: xdsl/dialects/experimental/stencil.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+class Stencil_Element(ParametrizedAttribute):
+    name = "stencil.element"
+    element = AnyOf([f32, f64])
+
+
+@dataclass(frozen=True)
+class Stencil_Index(ParametrizedAttribute):
+    # TODO: can you have an attr and an op with the same name?
+    name = "stencil.index"
+    shape = Annotated[ArrayAttr[IntAttr], ArrayLength(2)]
```
Comment (webmiche) [inline xdsl/dialects/experimental/stencil.py:117]: Why do you sometimes use this form and sometimes the `ParameterDef` form? We usually only use the later for ParametrizedAttributes, right?

**Review by webmiche — APPROVED** [review]
I am fine with merging this into the experimental folder as is. I just want to point out that you don't really have constructors, but I guess that can be your choice.

I think merging into experimental should be very liberal, so please go ahead. I don't quite like the core changes to ArrayAttr that are nested inside this PR. This makes the history quite weird and won't allow us to revert these changes in the future, but I am letting it slip for now.

File: xdsl/dialects/experimental/stencil.py (hunk 4 above)
Comment (AntonLydike) [reply xdsl/dialects/experimental/stencil.py:117]: Oh, I didn't even notice that we use `ParameterDef` elsewhere!

**Pull request merged**